File: Python/Programmers/Level 1/시저 암호.py
# def solution(s, n):
#     up, low = [], []
#     for i in range(ord("A"), ord("Z") + 1):
#         up.append(chr(i)) # 알파벳 대문자 리스트 
#         low.append(chr(i + 32)) # 알파벳 소문자 리스트
    
#     answer = ""
#     for word in s:
#         if word == " ":
#             answer += " " # 공백이면 공백 추가
#         # 알파벳 리스트 크기가 26이므로 z 다음 a가 나오게 하기 위해 26으로 나눈 나머지 값을 이용
#         elif ord("A") <= ord(word) <= ord("Z"):
#             # ord(word) - ord("A")는 word가 리스트에서 현재 몇 번째 인덱스인지,
#             # 여기서 +n을 하여 밀린 인덱스를 구하고 리스트 안에서만 값을 찾아야 하므로 26으로 나눔
#             answer += up[(ord(word) - ord("A") + n) % 26]
#         elif ord("a") <= ord(word) <= ord("z"):
#             answer += low[(ord(word) - ord("a") + n) % 26]
    
#     return answer

# def solution(s, n):
#     s = list(s) # s를 리스트로 변환
#     for i in range(len(s)):
#         if s[i].isupper(): # isupper() : 해당 문자가 대문자인지 체크
#             # 위 코드 처럼 알파벳 리스트를 따로 만들지 않고 밀린 인덱스를 구하고 +ord('A')를 하여 A로부터 밀린 암호 문자를 구함
#             # 즉, 아스키코드에서 알파벳 안에서만 암호 문자를 구해낼 수 있음
#             s[i]=chr((ord(s[i])-ord('A')+ n)%26+ord('A'))
#         elif s[i].islower(): # islower() : 해당 문자가 소문자인지 체크
#             s[i]=chr((ord(s[i])-ord('a')+ n)%26+ord('a'))

#     return "".join(s)

def solution(s, n):
    # s를 리스트로 변환하지 않고 변수를 두어 계산
    answer = ""
    for i in range(len(s)):
        if s[i].isupper():
            answer += chr((ord(s[i]) + n - ord("A")) % 26 + ord("A"))
        elif s[i].islower():
            answer += chr((ord(s[i]) + n - ord("a")) % 26 + ord("a"))
        else:
            answer += " "
    
    return answer

File: JavaScript/Programmers/Level 1/시저 암호.js
// 문자를 n만큼 뒤로 밀면 됨, "z"인 경우 1번 뒤로 밀면 "a"가 됨
// 알파벳은 26개 있으므로 % 연산자를 활용하여 "a" ~ "z", "A" ~ "Z"가 순환할 수 있도록 구현

// function solution(s, n) {
//     return [...s].map(alpha => { // 전개문법을 이용하여 문자열을 배열로 변환
//         const ascii = alpha.charCodeAt(0); // 현재 알파벳의 아스키 코드 값을 구함
//         let tmp = 0;
//         if ("A" <= alpha && alpha <= "Z") {
//             tmp = 65; // alpha가 대문자면 65
//         } else if ("a" <= alpha && alpha <= "z") {
//             tmp = 97; // alpha가 소문자면 97
//         }
        
//         // tmp가 0이라는 뜻은 공백이라는 뜻
//         // 삼항연산자를 이용하여 공백이면 공백 return, 알파벳이면 n만큼 뒤로 민 값에 해당하는 알파벳 return
//         return tmp ? String.fromCharCode((ascii - tmp + n) % 26 + tmp) : alpha;
//     }).join(""); // 배열을 다시 문자열로 join
// }

// 아스키 코드 값으로 변환하여 계산하지 않고 미리 알파벳 문자열을 선언 후 계산

// function solution(s, n) {
//     const upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"; // 대문자 문자열
//     const lower = "abcdefghijklmnopqrstuvwxyz"; // 소문자 문자열
    
//     return [...s].map(a => {
//         const alpha = upper.includes(a) ? upper : lower; // a가 대문자면 upper, 소문자면 lower 저장
//         // alpha에서 a의 인덱스가 -1이라는 뜻은 공백이라는 뜻
//         // 삼항 연산자를 이용하여 공백이면 공백 return, 공백이 아니면 n만큼 뒤로 민 알파벳 return
//         return alpha.indexOf(a) === -1 ? a : alpha[(alpha.indexOf(a) + n) % 26];
//     }).join(""); // 배열을 다시 문자열로 join
// }

// reduce를 사용한 풀이
function solution(s, n) {
    return [...s].reduce((answer, alpha) => {
        if (alpha === " ") {
            return answer + alpha;
        } else {
            if (alpha === alpha.toUpperCase()) {
                return answer + String.fromCharCode((alpha.charCodeAt() + n - 65) % 26 + 65);
            } else if (alpha === alpha.toLowerCase()) {
                return answer + String.fromCharCode((alpha.charCodeAt() + n - 97) % 26 + 97);
            }
        }
    }, "");
}
